### ISSUE TITLE
Could you please provide the trained weight file?

### ISSUE THREAD

**Issue by TianGao-NJUST** [Phuoc-Hoan-Le/BinaryViT#1]
Will too many residual branches make the main branch useless? Will the meaning of binarization disappear? For example, in the multi-head attention mechanism, after adding q, k, v, will its amplitude be much larger than the binary attention main branch? In FFN, multiple full-precision branches will be much larger than binary mlp. In this way, the model becomes a pool-former, but the order of channel interaction and token interaction is changed.

**Comment by Phuoc-Hoan-Le** [comment]
@812985728 Thank you for your questions and comments.

### "Could you please provide the trained weight file?":
Unfortunately, due to the company's policy, I cannot share with you the trained weights, however, I have set up the proper training configuration for all of the models that were recorded in the paper in the scripts folder.

### "Will too many residual branches make the main branch useless? Will the meaning of binarization disappear?"
Adding many residual branches doesn’t make the main branch useless as long as there is a scaling mechanism (such as Batch Normalization) right before the addition of the residual connection to make sure the main branch is properly scaled with respect to the residual connection. With extra residual connections, information loss from the sign function can be mitigated as mentioned in the ReActNet paper.

### "In FFN, multiple full-precision branches will be much larger than binary mlp."
Thank you for your insight. We have mainly added the multi-pooling branches to increase the representational capability without significantly increasing the number of OPs and without including convolutions. Also, we did not know there is a separate architecture called Pool-former. Thank you for letting us know about this.

If you have any other questions or comments please let us know

**Comment by TianGao-NJUST** [comment]
Thanks for the reply, this work is great and the performance exceeded my expectations. I'm just a little worried about whether self-attention is effective. The model thus becomes a pure mlp binary model. As far as I know, due to softmax, the binarization of self-attention will bring serious gradient disappearance and gradient errors. All of the above make the binary self-attention mechanism difficult to train. At the same time, it is difficult for binary attention to achieve full accuracy. Did the addition of q, k, v play an important role in the final result? With these three residual links, the gradient flow can bypass self-attention and directly reach the linear layer weights of q, k, v.

**Comment by Phuoc-Hoan-Le** [comment]
@812985728 Thank you for your comments on our work. I agree with you that the binarization of self-attention will bring serious gradient disappearance and gradient errors. For that reason, we use the method from BiT (https://arxiv.org/pdf/2205.13016.pdf), where the round function is applied after the softmax. Also, from BiT, a learnable scaling factor is used to divide the output of the softmax to balance out the number of 0's and 1's that will be outputted by the round function; and the same learnable scaling factor is then used to multiply the output of the round function to make that output have approximately the same scale as its full precision. However, this method is not without a flaw and there have been few recent works exploring how to improve binarizing the attention, especially for ViT. We just used BiT method, since that method was readily available during the development phase of this work. 

With regards to the addition of q, k, v, this was mainly based on ReActNet in which we wanted to preserve the information of q, k, v through the use of residual connections before their information gets lost by the sign function. This is not just for q, k ,v, but this also applies to the FC layers in the MLP layer and also applies to the output of the self-attention (the output of the matrix multiplication between the binarized V output and the binarized softmax output). Therefore, every fully connected layer including the self-attention will equally contribute to the final representational capability of the binary model. And therefore, the addition of q, k, v will play a role just as important as the addition of the output of the self-attention and the addition of the two FC layer outputs in the FFN.  We never thought of it as a helper with regards to the gradient flow to bypass self-attention and directly reach linear layer weights of q, k, v, but that's an interesting insight though.

**Comment by Phuoc-Hoan-Le** [comment]
@812985728 I will close this issue for now. If you have any more concerns, you are more than welcome to reopen this issue or email me.
